### PR TITLE
MAINT: Translate Digamma from Cython into C++

### PR DESCRIPTION
**Pull request by steppi**
#### Reference issue

This closes no issues, but is relevant to #19404 and #19481.
#### What does this implement/fix?

This PR translates `digamma.pxd` and its cephes dependencies `cephes/psi.c` and `cephes/zeta.c` into CUDA compatible C++ along the lines of #19471 and #19496

#### Additional information

This branch has been rebased upon the feature branch for the `loggamma` PR, #19496. The `loggamma` PR should be reviewed and merged first. I'm submitting this now in case @izaid would like a chance to review this after he reviews the `loggamma` PR but before the `loggamma` PR gets merged. I've again confirmed that this works in CuPy on Cuda. It might be time to think about a systematic way to test this, instead of relying on ad-hoc scripts.

### CODE REVIEW

**Comment by steppi** [comment]
Woah. I love the auto-labels. Thanks @lucascolley!

**Comment by ev-br** [comment]
>  It might be time to think about a systematic way to test this

CuPy has a handy decorator:

https://github.com/cupy/cupy/blob/main/tests/cupyx_tests/scipy_tests/special_tests/test_ellipk.py#L11

**Comment by steppi** [comment]
OK, all tests passed locally, but in

`Linux Meson tests / Build with gcc-8` and `Linux Meson tests / Prerelease deps, coverage, and 64-bit BLAS (3.9)(pull_request)` I'm seeing the identical failure

```
 =========================== short test summary info ============================
FAILED special/tests/test_mpmath.py::TestSystematic::test_digamma_complex - AssertionError: 
Max |adiff|: 0.000794359
Max |rdiff|: 0.000404781
Bad results (5 out of 439) for the following points (in output 0):
         0.-4.285714285714286j => 1.459849652666047-1.6874629934552177j != 1.4598496526660476-1.6874629934679093j  (rdiff          5.688000217407392e-12)
         0.-2.857142857142857j => 1.060164090818056-1.745796276593757j != 1.060164090818057-1.7457963769960356j  (rdiff          4.915687678361031e-08)
        0.-1.4285714285714286j => 0.4002307136439539-1.9203991474807345j != 0.4002307136439535-1.9211935061090597j  (rdiff           0.000404781189550142)
        0.+1.6666666666666667j => 0.5421986717787242+1.8707073573271982j != 0.542198671778724+1.8708852962625953j  (rdiff          9.135059536344818e-05)
        0.+3.3333333333333335j => 1.211543534548304+1.7207963242754367j != 1.211543534548303+1.7207963293143569j  (rdiff         2.3943394936997248e-09)
= 1 failed, 59615 passed, 2508 skipped, 260 xfailed, 13 xpassed, 6 warnings in 758.61s (0:12:38) =
```

That's a pretty bad tolerance violation. Not sure what's going on yet.

**Comment by steppi** [comment]
I've finished translating `wright_bessel` as well but will wait to PR that to avoid having too many open PRs at once.

https://github.com/steppi/scipy/compare/digamma_cpp...steppi:scipy:wright_bessel_cpp

**Comment by steppi** [comment]
That test was failing locally, I just didn't realize `python dev.py test -t scipy.special.tests` skips slow tests.  I should have been running `python dev.py test -t scipy.special.tests -m full`. The error was due to a typo in a condition.

File: scipy/special/special/constants.h [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#pragma once
+/* This header is for constants which are used in multiple places in this library */
+
+namespace special {
+namespace detail {
+    constexpr double MACHEP64 = 2.220446092504131E-16; // machine epsilon
```
[the remark this answers is not in the document]
Comment (steppi) [reply scipy/special/special/constants.h:6]: Great idea! Thanks for catching this oversight.  https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon

[the remark this answers is not in the document]
Comment (izaid) [reply scipy/special/special/constants.h]: Indeed, great idea! I think we'll find a lot of that as we go through cephes... We generally should be replacing those things, so long as we can do it without breakage.

**Review by izaid — APPROVED** [review]

**Comment by rlucas7** [comment]
> That test was failing locally, I just didn't realize `python dev.py test -t scipy.special.tests` skips slow tests. I should have been running `python dev.py test -t scipy.special.tests -m full`. The error was due to a typo in a condition.

This looks good to me when running `python dev.py test -t scipy.special.tests -m full`all tests pass except mpmath tests which still get skipped.

File: scipy/special/special/constants.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (rlucas7) [reply scipy/special/special/constants.h]: looks like these are updated in the tip of this PR so resolving.

**Review by rlucas7 — APPROVED** [review]
this LGTM. I'll merge this one after the loggamma PR later today if no further comments

**Comment by rlucas7** [comment]
@steppi it looks like there are conflicts now as a result of me merging https://github.com/scipy/scipy/pull/19496
could you rebase this PR off main and resolve any conflicts? 
I can merge once the conflicts are resolved.

**Comment by steppi** [comment]
The failure is unrelated

```
=========================== short test summary info ============================
FAILED scipy/optimize/tests/test__differential_evolution.py::TestDifferentialEvolutionSolver::test_immediate_updating - DeprecationWarning: This process (pid=16551) is multi-threaded, use of fork() may lead to deadlocks in the child.
= 1 failed, 45852 passed, 2368 skipped, 147 xfailed, 14 xpassed in 338.61s (0:05:38) =
```

**Comment by rlucas7** [comment]
> The failure is unrelated
> 
> ```
> =========================== short test summary info ============================
> FAILED scipy/optimize/tests/test__differential_evolution.py::TestDifferentialEvolutionSolver::test_immediate_updating - DeprecationWarning: This process (pid=16551) is multi-threaded, use of fork() may lead to deadlocks in the child.
> = 1 failed, 45852 passed, 2368 skipped, 147 xfailed, 14 xpassed in 338.61s (0:05:38) =
> ```

yeah I checked and it doesn't look like that test case is using digamma so this must be a flaky failure.  Merging, thanks for the rebase @steppi

**Pull request merged**